### PR TITLE
Add Sparsified sgd

### PR DESCRIPTION
**Pull request by negar-foroutan**
This adds an implementation of sparsifiedSGD (the sparse version of SGD). An optimizer class, a scheduler function,  and a communication function are added here.

### CODE REVIEW

File: mlbench/refimpls/pytorch/optim/optimizer.py [hunk 1]
```diff
@@ -15,16 +18,174 @@ def get_optimizer(options, model):
...
+        """
+        Sparsify the gradients vector by selecting 'k' of them randomly.
+        param model: learning model
+        param lr: learning rate
+        return: sparsified gradients vector ('k' gradients and their indices)
+        """
+        params_sparse_tensors = []
+
+        for ind, param in enumerate(model.parameters()):
+            self.state[param]['memory'] += param.grad.data * lr[0]
```
Comment (Panaetius) [inline mlbench/refimpls/pytorch/optim/optimizer.py]: If a layer or similar gets manually assigned a different tensor, the dict-key here would change (it's `id(parameter)` which maps to the python object identifier). This could be prevented by using `model.named_parameters()` which returns a unique name for each parameter along with the values. Then that name could be used as dict key.

File: mlbench/refimpls/pytorch/controlflow/controlflow.py [hunk 2]
```diff
@@ -32,9 +35,27 @@ def train_epoch(model, optimizer, criterion, scheduler, options, timeit):
...
+                                           options.random_sparse,
+                                           optimizer,
+                                           scheduler.get_lr())
+        else:
+            aggregate_gradients(model, options.world_size)
+
         optimizer.step()
 
+        if options.model_name == 'logistic_regression' and options.train_validate:
+            t = options.runtime['current_epoch'] * options.train_num_samples_per_device + batch_idx * options.batch_size
```
[the remark this answers is not in the document]
Comment (negar-foroutan) [reply mlbench/refimpls/pytorch/controlflow/controlflow.py]: I agree, I'll fix it.

File: mlbench/refimpls/pytorch/controlflow/controlflow.py [hunk 3]
```diff
@@ -46,6 +67,45 @@ def train_epoch(model, optimizer, criterion, scheduler, options, timeit):
...
+        if options.force_target_dtype:
+            target = convert_dtype(options.dtype, target)
+
+        if options.use_cuda:
+            data, target = data.cuda(), target.cuda()
+        target = target * 2 - 1
+
+        for weight in estimated_weights:
+            w = weight.squeeze()
+            batch_loss = np.log(1 + np.exp(-target * (data @ w)))
```
[the remark this answers is not in the document]
Comment (negar-foroutan) [reply mlbench/refimpls/pytorch/controlflow/controlflow.py]: Yes, I'll use the softmargin loss instead, thanks.

File: mlbench/refimpls/pytorch/controlflow/controlflow.py [hunk 4]
```diff
@@ -46,6 +67,45 @@ def train_epoch(model, optimizer, criterion, scheduler, options, timeit):
...
+            loss += batch_loss.sum().item()
+
+        num_samples += data.size()[0]
+
+    train_loss = global_average(loss, num_samples).item()
+
+    l2_loss = sum(weight.norm(2) ** 2 for weight in estimated_weights).item()
```
[the remark this answers is not in the document]
Comment (negar-foroutan) [reply mlbench/refimpls/pytorch/controlflow/controlflow.py]: As we just need to calculate L1 and L2 norm of a tensor here, I think using loss functions make it complicated.

File: mlbench/refimpls/pytorch/optim/optimizer.py [hunk 5]
```diff
@@ -15,16 +18,174 @@ def get_optimizer(options, model):
...
+        defaults = dict(lr=lr, weight_decay=weight_decay)
+
+        super(sparsified_SGD, self).__init__(params, defaults)
+
+        self.__create_gradients_memory()
+        self.__create_weighted_average_params()
+
+        self.num_coordinates = sparse_grad_size
+
+    def __setstate__(self, state):
```
[the remark this answers is not in the document]
Comment (negar-foroutan) [reply mlbench/refimpls/pytorch/optim/optimizer.py]: Yes, you're right. I forgot to remove it.

File: mlbench/refimpls/pytorch/optim/optimizer.py (hunk 1 above)
Comment (negar-foroutan) [reply mlbench/refimpls/pytorch/optim/optimizer.py]: I cannot use the name as the dict key, because the key here is not the name of the parameter, it's the parameter itself.
